unit_tests/test_redis_client_base.py

```diff
@@ -0,0 +1,235 @@
+                           match="API unregister process timed out. "
+                                 "Server will continue processing "
+                                 "your request on the background and "
+                                 "will notify you as soon as "
+                                 "the unregistration has been completed."):
+            redis_client_auto_register.is_active = True
+            redis_client_auto_register.unregister(is_blocking=True)
+        mock_wait_until_timeout_blocking.assert_called()
+
+    @staticmethod
+    @patch("uuid.uuid4", return_value="some-transaction-uuid")
+    def test_on_register(mock_uuid, redis_client_auto_register):
+        """Check the on_register function with correct message that doesn't throw exception."""
+        redis_client_auto_register.register(is_blocking=False)
+        data = {"name": AREA_ID, "device_uuid": DEVICE_ID, "transaction_id": mock_uuid()}
+        message = {"data": json.dumps(data)}
+        redis_client_auto_register._on_register(message)
+        assert redis_client_auto_register.is_active is True
+        assert redis_client_auto_register.area_uuid == DEVICE_ID
+
+    @staticmethod
+    @patch("uuid.uuid4", return_value="some-transaction-uuid")
+    def test_on_unregister(mock_uuid, redis_client_auto_register):
+        """Check the on_unregister function with correct message that doesn't throws exception."""
+        redis_client_auto_register.is_active = True
+        redis_client_auto_register.unregister(is_blocking=False)
+        data = {"name": AREA_ID, "device_id": DEVICE_ID,
+                "transaction_id": mock_uuid(), "response": "success"}
+        message = {"data": json.dumps(data)}
+        redis_client_auto_register._on_unregister(message)
+        assert redis_client_auto_register.is_active is False
+
+    @staticmethod
+    @patch("uuid.uuid4", return_value="some-transaction-uuid")
+    def test_on_unregister_throws_exception(mock_uuid, redis_client_auto_register):
+        """Check if exception is raised when response of on_unregister is not successful."""
+        with pytest.raises(RedisAPIException,
+                           match=f"Failed to unregister from market {AREA_ID}."
+                                 "Deactivating connection."):
+            redis_client_auto_register.is_active = True
+            redis_client_auto_register.unregister(is_blocking=False)
+            data = {"name": AREA_ID, "device_id": DEVICE_ID,
+                    "transaction_id": mock_uuid(), "response": "unsuccessful"}
+            message = {"data": json.dumps(data)}
+            redis_client_auto_register._on_unregister(message)
+
+    @staticmethod
+    def test_select_aggregator(redis_client_auto_register):
+        """Check if redis_db.publish is called with correct arguments to select aggregator."""
+        aggregator_uuid = str(uuid.uuid4())
+        redis_client_auto_register.redis_db.publish = MagicMock()
```

Comment:
Nit, what is the scope of these mocks? are they reset after the test run ?

Reply:
That's a good question. As far as I see from the docs, the `patch` decorators/context managers are
> used for patching objects only within the scope of the function they decorate. They automatically handle the unpatching for you, even if exceptions are raised.

In this case, we're manually adding the mock, so I assume that it will stay there for the entire life of the `redis_client_auto_register` fixture object. However, the default scope of a fixture in pytest is just `"function"`, and it will be destroyed once the function returns: https://docs.pytest.org/en/latest/how-to/fixtures.html#fixture-scopes

So it shouldn't make a difference in this case.

Reply:
Good to know, thanks!